File: numericalFunctions/ptwXY/Test/UnitTesting/convolution/gaussian.c
/*
# <<BEGIN-copyright>>
# Copyright 2022, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>
*/

#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include <ptwXY.h>
#include <nf_utilities.h>
#include <nfut_utilities.h>
#include <ptwXY_utilities.h>

static int verbose = 0;
static char *fmtXY = "%.12e %.12e\n";

static void printIfVerbose( const char * const label, ptwXYPoints *data, double accuracy, double offset, double sigma, double amplitude, 
    double xMin, double xMax, double dullEps );
/*
************************************************************
*/
int main( int argc, char **argv ) {

    int iarg, echo = 0;
    double accuracy = 1e-5, area, offset, amplitude = 5., sigma = 10., xMin, xMax;
    ptwXYPoints *gaussian;
    nfu_status status;
    statusMessageReporting smr;

    smr_initialize( &smr, smr_status_Ok );

    for( iarg = 1; iarg < argc; iarg++ ) {
        if( strcmp( "-v", argv[iarg] ) == 0 ) {
            verbose = 1; }
        else if( strcmp( "-e", argv[iarg] ) == 0 ) {
            echo = 1; }
        else {
            nfu_printErrorMsg( "ERROR %s: invalid input option '%s'", __FILE__, argv[iarg] );
        }
    }
    if( echo ) printf( "%s\n", __FILE__ );

    if( ( gaussian = ptwXY_createGaussianCenteredSigma1( &smr, accuracy ) ) == NULL ) 
        nfut_printSMRErrorExit2p( &smr, "Via." );
    printIfVerbose( "ptwXY_createGaussianCenteredSigma1", gaussian, accuracy, 0., 1., 1., 0., 0., 0. );

    if( ( status = ptwXY_integrateDomain( &smr, gaussian, &area ) ) != nfu_Okay ) nfut_printSMRErrorExit2p( &smr, "Via." );
    if( verbose ) printf( "# area = %.12e  area err = %e\n", area, area / 2.50662827463 - 1 );
    if( fabs( area / 2.50662827463 - 1 ) > 0.2 * accuracy )
        nfu_printErrorMsg( "ERROR %s: area err = %e, area = %.12e", __FILE__, area / 2.50662827463 - 1, area );
    ptwXY_free( gaussian );

    offset = 110.;
    xMin = offset - 66.;
    xMax = offset + 66.;
    if( ( gaussian = ptwXY_createGaussian( &smr, accuracy, offset, sigma, amplitude, xMin, xMax, 0. ) ) == NULL )
        nfut_printSMRErrorExit2p( &smr, "Via." );
    printIfVerbose( "ptwXY_createGaussian", gaussian, accuracy, offset, sigma, amplitude, xMin, xMax, 0. );
    if( ( status = ptwXY_integrateDomain( &smr, gaussian, &area ) ) != nfu_Okay ) nfut_printSMRErrorExit2p( &smr, "Via." );
    if( verbose ) printf( "# area = %.12e  area err = %e\n", area, area / ( sigma * amplitude * 2.50662827463 ) - 1 );
    if( fabs( area / ( sigma * amplitude * 2.50662827463 ) - 1 ) > 0.2 * accuracy )
        nfu_printErrorMsg( "ERROR %s: area err = %e, area = %.12e", __FILE__, area / ( sigma * amplitude * 2.50662827463 ) - 1, area );
    ptwXY_free( gaussian );

    return( 0 );
}
/*
************************************************************
*/
static void printIfVerbose( const char * const label, ptwXYPoints *data, double accuracy, double offset, double sigma, double amplitude, 
    double xMin, double xMax, double dullEps ) {

    if( !verbose ) return;
    printf( "# %s\n", label );
    printf( "# accuracy = %g\n", accuracy );
    printf( "# offset = %g\n", offset );
    printf( "# sigma = %g\n", sigma );
    printf( "# amplitude = %g\n", amplitude );
    printf( "# xMin = %.17e\n", xMin );
    printf( "# xMax = %.17e\n", xMax );
    printf( "# dullEps = %g\n", dullEps );
    printf( "# length = %d\n", (int) data->length );
    ptwXY_simpleWrite( data, stdout, fmtXY );
    printf( "\n\n" );
}

File: numericalFunctions/ptwXY/Python/Test/UnitTesting/convolution/gaussian.py
# <<BEGIN-copyright>>
# Copyright 2021, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: BSD-3-Clause
# <<END-copyright>>

import os

from numericalFunctions import pointwiseXY_C

if( 'CHECKOPTIONS' in os.environ ) :
    options = os.environ['CHECKOPTIONS'].split( )
    if( '-e' in options ) : print( __file__ )

CPATH = '../../../../Test/UnitTesting/convolution'

os.system( 'cd %s; ./gaussian -v > v' % CPATH )
f = open( os.path.join( CPATH, 'v' ) )
ls = f.readlines( )
f.close( )

def getDoubleValue( name, ls ) :

    s = "# %s = " % name
    n = len( s )
    if( ls[0][:n] != s ) : raise Exception( '%s: line does not contain %s info: "%s"' % ( __file__, name, ls[0][:-1] ) )
    value = float( ls[0].split( '=' )[1] )
    return( ls[1:], value )

def compareValues( label, i, v1, v2 ) :

    sv1, sv2 = '%.12g' % v1, '%.12g' % v2
    sv1, sv2 = '%.8g' % float( sv1 ), '%.8g' % float( sv2 )
    if( sv1 != sv2 ) : print( '<%s> <%s>' % ( sv1, sv2 ) )
    if( sv1 != sv2 ) : raise Exception( '%s: values %s %s diff by %g at %d for label = %s' % ( __file__, v1, v2, v2 - v1, i, label ) )

def getData( ls ) :

    i = 0
    for l in ls :
        if( l.strip( ) != '' ) : break
        i = i + 1
    ls = ls[i:]
    if( len( ls ) == 0 ) : return( ls )
    if( ls[0][:9] == "# area = " ) : ls = ls[1:]
    if( len( ls ) == 0 ) : return( ls )

    label = ls[0][2:-1]
    if( label not in [ 'ptwXY_createGaussianCenteredSigma1', 'ptwXY_createGaussian' ] ) :
        raise Exception( '%s: invalid label = "%s"' % ( __file__, ls[0][:-1] ) )
    ls = ls[1:]
    ls, accuracy = getDoubleValue( 'accuracy', ls )
    ls, offset = getDoubleValue( 'offset', ls )
    ls, sigma = getDoubleValue( 'sigma', ls )
    ls, amplitude = getDoubleValue( 'amplitude', ls )
    ls, xMin = getDoubleValue( 'xMin', ls )
    ls, xMax = getDoubleValue( 'xMax', ls )
    ls, dullEps = getDoubleValue( 'dullEps', ls )

    length, ls = ls[0], ls[1:]
    if( '# length = ' != length[:11] ) : raise Exception( '%s: line does not contain length info: "%s"' % ( __file__, length.strip( ) ) )
    length = int( length.split( '=' )[1] )

    data = [ list( map( float, ls[i].split( )[:2] ) ) for i in range( length ) ]
    data = pointwiseXY_C.pointwiseXY_C( data, initialSize = len( data ), overflowSize = 10 )
    ls = ls[length:]
    if( label == 'ptwXY_createGaussianCenteredSigma1' ) :
        gaussian = pointwiseXY_C.basicGaussian( accuracy )
    else :
        gaussian = pointwiseXY_C.gaussian( accuracy = accuracy, domainMin = xMin, domainMax = xMax, offset = offset, sigma = sigma, 
            amplitude = amplitude, dullEps = dullEps )
    if( len( data ) != len( gaussian ) ) : raise Exception( '%s: len( data ) = %d != len( gaussian ) = %d for label "%s"' %
        ( __file__, len( data ), len( gaussian ), label ) )
    for i , dXY in enumerate( data ) :
        gXY = gaussian[i]
        compareValues( label, i, dXY[0], gXY[0] )
        compareValues( label, i, dXY[1], gXY[1] )
    return( ls )

while( len( ls ) ) : ls = getData( ls )
